Error early when dataclass is not registered

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112211


Partially fixes: https://github.com/pytorch/pytorch/issues/112043
cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @avikchaudhuri @gmagogsfm @zhxchen17 @angelayi @suo

tugsbayasgalan committed:
Error early when dataclass is not registered

[ghstack-poisoned]

zhxchen17 commented on test/dynamo/test_export.py:
```diff
@@ -4238,6 +4239,25 @@ def forward(self, x):
+
+        with self.assertRaisesRegex(
+            RuntimeError,
+            "Perhaps you forgot to register it as dataclass. "
+            "Please take a look at torch._export.utils.register_dataclass_as_pytree_node.",
```
```suggestion
            "Dataclasses are supposed to be pytree nodes to be exportable."
            "Please take a look at torch._export.utils.register_dataclass_as_pytree_node for more information.",
```

tugsbayasgalan committed:
Update test/dynamo/test_export.py

Co-authored-by: Zhengxu Chen <zhxchen17@outlook.com>

tugsbayasgalan committed:
Update on "Error early when dataclass is not registered"



Partially fixes: https://github.com/pytorch/pytorch/issues/112043
cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng avikchaudhuri gmagogsfm zhxchen17 angelayi

[ghstack-poisoned]

suo commented on torch/_dynamo/eval_frame.py:
```diff
@@ -1201,6 +1202,22 @@ def result_capturing_wrapper(*graph_inputs):
+                if dataclasses.is_dataclass(arg):
+                    raise RuntimeError(
+                        "Dataclasses are supposed to be pytree nodes to be exportable."
+                        "Please take a look at torch._export.utils.register_dataclass_as_pytree_node "
```
depressing that we're suggesting private APIs to people

zhxchen17 replied:
yeah actually we should just publish this API. @tugsbayasgalan

angelayi replied:
It is public [`torch.export.register_dataclass`](https://fburl.com/code/hg2la1lc)

tugsbayasgalan committed:
Update on "Error early when dataclass is not registered"



Partially fixes: https://github.com/pytorch/pytorch/issues/112043
cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng avikchaudhuri gmagogsfm zhxchen17 angelayi

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Error early when dataclass is not registered"



Partially fixes: https://github.com/pytorch/pytorch/issues/112043
cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng avikchaudhuri gmagogsfm zhxchen17 angelayi

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Error early when dataclass is not registered"



Partially fixes: https://github.com/pytorch/pytorch/issues/112043
cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng avikchaudhuri gmagogsfm zhxchen17 angelayi

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Error early when dataclass is not registered"



Partially fixes: https://github.com/pytorch/pytorch/issues/112043
cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng avikchaudhuri gmagogsfm zhxchen17 angelayi

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge -f "Only previous failure was lint error"

ZainRizvi commented:
@pytorchbot revert -m "Breaks internal builds. See D50820325" -c ghfirst

ZainRizvi commented:
@tugsbayasgalan PTAL
